### ISSUE TITLE
[Enhancement] Update to Crystal 1.0.0

### ISSUE THREAD

**Issue by TheFrenchGhosty** [iv-org/invidious#1933]
**Is your enhancement request related to a problem? Please describe.**

Crystal 1.0.0 came out: https://github.com/crystal-lang/crystal/releases/tag/1.0.0

Invidious needs to be tested, and if stuff are broken, fixed to be working on Crystal 1.0.0.

**Describe the solution you'd like**

Invidious must work on Crystal 1.0.0, previous versions don't matter.

**Describe alternatives you've considered**

N/A

**Additional context**

N/A

**Comment by shimuldn** [comment]
Instalation fali

```
✔ Finished Database section
➜ Updating config.yml with new info...
sed: -e expression #1, char 110: unknown option to `s'
Resolving dependencies
Fetching https://github.com/will/crystal-pg.git
Fetching https://github.com/crystal-lang/crystal-sqlite3.git
Fetching https://github.com/kemalcr/kemal.git
Fetching https://github.com/ysbaddaden/pool.git
Fetching https://github.com/omarroth/protodec.git
Fetching https://github.com/iv-org/lsquic.cr.git
Fetching https://github.com/crystal-lang/crystal-db.git
Fetching https://github.com/luislavena/radix.git
Unable to satisfy the following requirements:

- `crystal (>= 0.35.0, < 2.0.0)` required by `pg 0.23.2`
- `crystal (>= 0.35.0, < 2.0.0)` required by `sqlite3 0.18.0`
- `crystal (~> 0.35, >= 0.35.0)` required by `kemal 0.27.0`
- `crystal (< 1.0.0)` required by `pool 0.2.3`
- `crystal (~> 0.34, >= 0.34.0)` required by `protodec 0.1.3`
- `crystal (~> 0.35, >= 0.35.1)` required by `lsquic 2.18.1-1`
- `crystal (>= 0.35.0, < 2.0.0)` required by `db 0.10.1`
- `crystal (< 1.0.0)` required by `radix 0.3.9`
Failed to resolve dependencies, try updating incompatible shards or use --ignore-crystal-version as a workaround if no update is available.
Showing last frame. Use --error-trace for full trace.

In src/invidious.cr:19:1

 19 | require "kemal"
      ^
Error: can't find file 'kemal'

If you're trying to require a shard:
- Did you remember to run `shards install`?
- Did you make sure you're running the compiler in the same directory as your shard.yml?

✗[ERROR] Process Failed!
```

**Comment by Perflyst** [comment]
Some deprecation warnings with crystal 1.0:

```
In src/invidious/jobs/bypass_captcha_job.cr:115:37

 115 | cookies = HTTP::Cookies.from_headers(headers)
                               ^-----------
Warning: Deprecated HTTP::Cookies.from_headers. Use `.from_client_headers` or `.from_server_headers` instead.

In /usr/share/crystal/src/http/cookie.cr:207:35

 207 | new.tap { |cookies| cookies.fill_from_headers(headers) }
                                   ^----------------
Warning: Deprecated HTTP::Cookies#fill_from_headers. Use `#fill_from_client_headers` or `#fill_from_server_headers` instead.

In src/invidious/routes/login.cr:241:33

 241 | cookies = HTTP::Cookies.from_headers(headers)
                               ^-----------
Warning: Deprecated HTTP::Cookies.from_headers. Use `.from_client_headers` or `.from_server_headers` instead.

In src/invidious/routes/login.cr:264:33

 264 | cookies = HTTP::Cookies.from_headers(headers)
                               ^-----------
Warning: Deprecated HTTP::Cookies.from_headers. Use `.from_client_headers` or `.from_server_headers` instead.

In src/invidious.cr:443:29

 443 | cookies = HTTP::Cookies.from_headers(headers)
                               ^-----------
Warning: Deprecated HTTP::Cookies.from_headers. Use `.from_client_headers` or `.from_server_headers` instead.

In src/invidious/users.cr:463:27

 463 | cookies = HTTP::Cookies.from_headers(headers)
                               ^-----------
Warning: Deprecated HTTP::Cookies.from_headers. Use `.from_client_headers` or `.from_server_headers` instead.

A total of 6 warnings were found.

**Comment by diolwe** [comment]
Related?

As of today `shards update` throws

````
Unable to satisfy the following requirements:

- `crystal (>= 0.35.0, < 2.0.0)` required by `pg 0.23.2`
- `crystal (>= 0.35.0, < 2.0.0)` required by `sqlite3 0.18.0`
- `crystal (~> 0.35, >= 0.35.0)` required by `kemal 0.27.0`
- `crystal (*)` required by `pool 0.2.4`
- `crystal (~> 0.34, >= 0.34.0)` required by `protodec 0.1.3`
- `crystal (~> 0.35, >= 0.35.1)` required by `lsquic 2.18.1-1`
- `crystal (>= 0.35.0, < 2.0.0)` required by `db 0.10.1`
- `crystal (< 1.0.0)` required by `radix 0.3.9`
Failed to resolve dependencies, try updating incompatible shards or use --ignore-crystal-version as a workaround if no update is available.
````
and `shards install`  throws 

````
Unable to satisfy the following requirements:

- `crystal (>= 0.35.0, < 2.0.0)` required by `pg 0.23.2`
- `crystal (>= 0.35.0, < 2.0.0)` required by `sqlite3 0.18.0`
- `crystal (~> 0.35, >= 0.35.0)` required by `kemal 0.27.0`
- `crystal (*)` required by `pool 0.2.4`
- `crystal (~> 0.34, >= 0.34.0)` required by `protodec 0.1.3`
- `crystal (~> 0.35, >= 0.35.1)` required by `lsquic 2.18.1-1`
- `crystal (>= 0.35.0, < 2.0.0)` required by `db 0.10.1`
- `crystal (< 1.0.0)` required by `radix 0.3.9`
Failed to resolve dependencies, try updating incompatible shards or use --ignore-crystal-version as a workaround if no update is available.
````
If I use `--ignore-crystal-version` in both cases `install` tells me:

````
Resolving dependencies
Fetching https://github.com/will/crystal-pg.git
Fetching https://github.com/crystal-lang/crystal-db.git
Fetching https://github.com/crystal-lang/crystal-sqlite3.git
Fetching https://github.com/kemalcr/kemal.git
Fetching https://github.com/luislavena/radix.git
Fetching https://github.com/jeromegn/kilt.git
Fetching https://github.com/crystal-loot/exception_page.git
Fetching https://github.com/ysbaddaden/pool.git
Fetching https://github.com/omarroth/protodec.git
Fetching https://github.com/iv-org/lsquic.cr.git
Shard "lsquic" version (2.18.1) doesn't match tag version (2.18.1-1)
Using db (0.10.1)
Using pg (0.23.2)
Using sqlite3 (0.18.0)
Using radix (0.3.9)
Using kilt (0.4.1)
Using exception_page (0.1.5)
Using kemal (0.27.0)
Using pool (0.2.4)
Using protodec (0.1.3)
Using lsquic (2.18.1-1)
Shard "radix" may be incompatible with Crystal 1.0.0
Shard "kemal" may be incompatible with Crystal 1.0.0
Shard "protodec" may be incompatible with Crystal 1.0.0
Shard "lsquic" may be incompatible with Crystal 1.0.0
````
I ignored the "may be incompatible" and went on with `crystal build src/invidious.cr --release`
I got:
````
In src/invidious/jobs/bypass_captcha_job.cr:115:37

 115 | cookies = HTTP::Cookies.from_headers(headers)
                               ^-----------
Warning: Deprecated HTTP::Cookies.from_headers. Use `.from_client_headers` or `.from_server_headers` instead.

In /usr/share/crystal/src/http/cookie.cr:207:35

 207 | new.tap { |cookies| cookies.fill_from_headers(headers) }
                                   ^----------------
Warning: Deprecated HTTP::Cookies#fill_from_headers. Use `#fill_from_client_headers` or `#fill_from_server_headers` instead.

In src/invidious/routes/login.cr:241:33

 241 | cookies = HTTP::Cookies.from_headers(headers)
                               ^-----------
Warning: Deprecated HTTP::Cookies.from_headers. Use `.from_client_headers` or `.from_server_headers` instead.

In src/invidious/routes/login.cr:264:33

 264 | cookies = HTTP::Cookies.from_headers(headers)
                               ^-----------
Warning: Deprecated HTTP::Cookies.from_headers. Use `.from_client_headers` or `.from_server_headers` instead.

In src/invidious.cr:454:29

 454 | cookies = HTTP::Cookies.from_headers(headers)
                               ^-----------
Warning: Deprecated HTTP::Cookies.from_headers. Use `.from_client_headers` or `.from_server_headers` instead.

In src/invidious/users.cr:463:27

 463 | cookies = HTTP::Cookies.from_headers(headers)
                               ^-----------
Warning: Deprecated HTTP::Cookies.from_headers. Use `.from_client_headers` or `.from_server_headers` instead.

A total of 6 warnings were found.
````
I ended up with a "kind of working"-invidious version, which wrongly tells me that all my channels do not not exist if I click on the channel name. Page footer contains ` Current version: 2021.03.13-502e8da @ master `

**Comment by shimuldn** [comment]
@diolwe 
`apt install crystal=0.36.1-1` will fix the issue for now.
You can use this [script](https://github.com/tmiland/Invidious-Updater/blob/master/invidious_update.sh) if you like.

**Comment by diolwe** [comment]
> `apt install crystal=0.36.1-1` will fix the issue for now.

Thank you! That lead to a working version of invidious - including the channel options.

**Comment by diolwe** [comment]
Due to changes on the side of crystal maintainers and their repo: older crystal versions are not available anymore? Crystal seems to be stuck with Crystal 1.0.0 [dd40a2442] (2021-03-22)

shards update && shards install therefore leads to the before-mentioned errors.

**Comment by Perflyst** [comment]
You can use the deb package or binary from https://github.com/crystal-lang/crystal/releases/tag/0.36.1

**Comment by diolwe** [comment]
@Perflyst Thank you! Compiles without errors now.

Shouldn't https://docs.invidious.io/Installation.md consider that as the default, mention the correct crystal version and link to the crystal release page on github?

**Comment by unixfox** [comment]
I have been running my Invidious instance (https://yewtu.be) under Crystal 1.0.0 for almost 2 days now, and it is running fine like it would with an older Crystal version.

Can we move forward with Crystal 1.0.0 on Invidious and support it? We just need to:
- Support Crystal 1.0.0 on lsquic
- Then merge this PR: https://github.com/iv-org/invidious/pull/2115

@SamantazFox @syeopite Could you work on this? I'm really not an expert in shards.

**Comment by syeopite** [comment]
Closed by #2115
